### PR TITLE
User story 5 create a resource

### PR DESCRIPTION
**Pull request by robotscissors**
Created document controller, views and model.
Added the user_id to the document table and restricted deletions to admins and document owners.

Also, added routing so that logged in users will get routed to a dashboard and non-authenticated users will be directed to the about page.

### CODE REVIEW

File: app/controllers/documents_controller.rb [hunk 1]
```diff
@@ -0,0 +1,69 @@
+class DocumentsController < ApplicationController
+  def index
+      @documents = Document.all.order(created_at: :desc)
```
Comment (alexstophel) [inline app/controllers/documents_controller.rb:3]: Indentation is off here (too many spaces)

File: app/controllers/documents_controller.rb [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+  end
+
+  def new
+    @document = Document.new
+  end
+
+  def create
+    @document = Document.new
+    @document.title = params[:document][:title]
```
Comment (alexstophel) [inline app/controllers/documents_controller.rb:12]: Consider using strong params! `params.require(:document).permit(:title, :description, :keywords)` could be extracted into a method and then `@document = Document.new(method_name)` would work here.

File: app/controllers/documents_controller.rb [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    @document.description = params[:document][:description]
+    @document.keywords = params[:document][:keywords]
+    @document.user_id = current_user.id
```
Comment (alexstophel) [inline app/controllers/documents_controller.rb:15]: Consider instantiating the new document off of the current_user object to get this line for free: `current_user.documents.build(params)`

File: app/controllers/documents_controller.rb [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    @document = Document.find(params[:id])
+  end
+
+  def edit
+    @document = Document.find(params[:id])
+  end
+
+  def update
+    @document = Document.find(params[:id])
+    @document.assign_attributes(document_params)
```
Comment (alexstophel) [inline app/controllers/documents_controller.rb:37]: You could remove this line and change the if statement to `if @document.update(document_params)`. Update returns a boolean value.

File: app/controllers/documents_controller.rb [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    end
+
+  end
+
+
+
+  private
+  # Using a private method to encapsulate the permissible parameters
+  def document_params
+    params.require(:document).permit(:title, :description, :keywords)
```
Comment (alexstophel) [inline app/controllers/documents_controller.rb:66]: Nice! This is what I was talking about above. Use this in create.

File: app/models/document.rb [hunk 6]
```diff
@@ -0,0 +1,7 @@
+class Document < ActiveRecord::Base
+  validates :title, length: {minimum: 5}, presence: true
+  validates :description, length: {minimum: 20}, presence: true
+  validates :user_id, presence: true
+  belongs_to :user
+
```
Comment (alexstophel) [inline app/models/document.rb:6]: Useless newline.

File: app/views/documents/_form.html.erb [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+  <strong> Title of Document </strong> <%= f.text_field :title, class: 'form-control', placeholder: "Enter document title (min. 5 characters)", autofocus: true %>
+<% end %>
+<%= form_group_tag(document.errors[:description]) do %>
+  <strong> Description </strong> <%= f.text_area :description, cols: 30, rows: 5, class: 'form-control', placeholder: "Enter document title (min. 20 characters)" %>
+<% end %>
+<%= form_group_tag(document.errors[:keywords]) do %>
+  <strong> Any Keywords That Apply </strong>&nbsp&nbsp&nbsp<small>(separate keywords by a space)</small><%= f.text_area :keywords, cols: 30, rows: 5,class: 'form-control', placeholder: "(optional)" %>
+<% end %>
+
+
```
Comment (alexstophel) [inline app/views/documents/_form.html.erb:22]: Useless newline.

File: app/views/documents/edit.html.erb [hunk 8]
```diff
@@ -0,0 +1,12 @@
+<h2>Edit Document</h2>
+<div class="row">
+  <div class="col-md-5 jumbotron">
+    <p>Some quick form instructions</p>
+
+      <%= render partial: "form", locals: { document: @document } %>
```
Comment (alexstophel) [inline app/views/documents/edit.html.erb:6]: Spacing seems off in this file.

File: app/views/documents/new.html.erb [hunk 9]
```diff
@@ -0,0 +1,12 @@
+<h2>Upload a New Document</h2>
+<div class="row">
+  <div class="col-md-5 jumbotron">
+    <p>Some quick form instructions</p>
+
+      <%= render partial: "form", locals: { document: @document } %>
```
Comment (alexstophel) [inline app/views/documents/new.html.erb:6]: Spacing seems off in this file.

File: app/views/documents/show.html.erb [hunk 10]
```diff
@@ -0,0 +1,25 @@
+<div class="row">
+  <div class="col-md-5 jumbotron">
+
```
Comment (alexstophel) [inline app/views/documents/show.html.erb:3]: Useless newline.

File: app/views/documents/show.html.erb [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+      <br /><br />
+      <small><%= "Original: #{@document.created_at.localtime}"%><br /><%="Modified: #{@document.updated_at.localtime}" %></small>
+
+      <div class="pull-right margin-before-50">
+           <%= link_to "Edit Document", edit_document_path, class: 'btn btn-success' %>
+           <%= if (current_user.admin? || (current_user.id === @document.user_id))
+             link_to "Delete", @document, method: :delete, class: 'btn btn-danger', data: { confirm: 'Are you sure you want to delete this post?' }
+           end %>
+      </div>
+
```
Comment (alexstophel) [inline app/views/documents/show.html.erb:18]: 2 useless newlines.

File: config/routes.rb [hunk 12]
```diff
@@ -1,14 +1,18 @@
 Rails.application.routes.draw do
+
```
Comment (alexstophel) [inline config/routes.rb:2]: Remove these 3 newlines.

File: spec/factories/documents.rb [hunk 13]
```diff
@@ -0,0 +1,8 @@
+FactoryGirl.define do
+  factory :document do
```
Comment (alexstophel) [inline spec/factories/documents.rb:2]: Consider adding Faker and making these values more realistic (and unique each time).

**Review by alexstophel — APPROVED** [review]
This looks awesome!

File: app/controllers/documents_controller.rb (hunk 1 above)
Comment (robotscissors) [reply app/controllers/documents_controller.rb:3]: fixed.

File: app/controllers/documents_controller.rb (hunk 2 above)
Comment (robotscissors) [reply app/controllers/documents_controller.rb:12]: Done.

File: app/controllers/documents_controller.rb (hunk 3 above)
Comment (robotscissors) [reply app/controllers/documents_controller.rb:15]: I think this is part magic.  I reduced 5 lines to this:
if current_user.documents.build(document_params).save

Can we talk about this?

File: app/models/document.rb (hunk 6 above)
Comment (robotscissors) [reply app/models/document.rb:6]: done.

File: app/views/documents/new.html.erb (hunk 9 above)
Comment (robotscissors) [reply app/views/documents/new.html.erb:6]: Done.

File: app/views/documents/edit.html.erb (hunk 8 above)
Comment (robotscissors) [reply app/views/documents/edit.html.erb:6]: done.

File: app/views/documents/show.html.erb (hunk 10 above)
Comment (robotscissors) [reply app/views/documents/show.html.erb:3]: fixed.

File: app/views/documents/show.html.erb (hunk 11 above)
Comment (robotscissors) [reply app/views/documents/show.html.erb:18]: fixed

File: spec/factories/documents.rb (hunk 13 above)
Comment (robotscissors) [reply spec/factories/documents.rb:2]: Good advice. Done.

File: config/routes.rb (hunk 12 above)
Comment (robotscissors) [reply config/routes.rb:2]: Done.

**Comment by robotscissors** [comment]
Closing and merging pull.

**Pull request merged**